wallet-api/src/Wallet/Emulator/Http.hs

```diff
@@ -152,16 +180,67 @@ runM state r = do
   res <- liftIO . runExceptT . flip runReaderT state . unM $ r
   liftEither res
 
-walletHandlers :: State -> Server WalletAPI
-walletHandlers state =
-  hoistServer walletApi (runM state) $
-  wallets :<|> fetchWallet :<|> createWallet :<|> createPaymentWithChange :<|>
-  payToPublicKey :<|>
-  submitTxn :<|>
-  getTransactions
+walletHandlers :: State -> Server API
+walletHandlers state = hoistServer api (runM state) $ walletApi :<|> controlApi
+  where
+    walletApi =
+      wallets :<|> fetchWallet :<|> createWallet :<|> createPaymentWithChange :<|>
+      payToPublicKey :<|>
+      submitTxn :<|>
+      getTransactions
+    controlApi =
+      blockchainActions :<|> setValidationData :<|> blockValidated :<|>
+      blockHeight
+
+handleNotifications ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Wallet
+  -> [Notification]
+  -> m ()
+handleNotifications = runWalletApiAction Types.handleNotifications
 
-walletApi :: Proxy WalletAPI
-walletApi = Proxy
+blockValidated ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Wallet
+  -> Block
+  -> m ()
+blockValidated wallet block = handleNotifications wallet [BlockValidated block]
+
+blockHeight ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Wallet
+  -> Height
+  -> m ()
+blockHeight wallet height = handleNotifications wallet [BlockHeight height]
+
+setValidationData :: (MonadReader State m, MonadIO m) => ValidationData -> m ()
+setValidationData vd = do
+  var <- asks getState
+  liftIO . atomically $ modifyTVar var (set validationData vd)
+
+blockchainActions :: (MonadReader State m, MonadIO m) => m [Tx]
+blockchainActions = do
+  var <- asks getState
+  liftIO . atomically $ blockchainActionsSTM var
+
+blockchainActionsSTM :: TVar EmulatorState -> STM [Tx]
+blockchainActionsSTM var = do
+  es <- readTVar var
+  let processed = validateEm es <$> emTxPool es
+      validated = catMaybes processed
+      block = validated
+      newState = addBlock block . emptyPool $ es
+  writeTVar var newState
+  pure block
+  where
+    addBlock block es = es & chain %~ (:) block
+    emptyPool es = es & txPool .~ []
```

Comment:
This can just be `emptyPool es = es { _txPool = [] }`.

Reply:
Tastes differ. I like `es & txPool .~ []` much more.

Reply:
I like lenses but I don't like the glyphs. (Because I know you all wanted another 2c added to this. :-D )

Reply:
No, can't resist. Aren't these more readable than the squiggles?

``` haskell
  addBlock block es = over chain (:) block es
  emptyPool es = set txPool [] es
```

_eta reduce to taste_

Reply:
ok, I'm not getting involved in a style issue. If someone *strongly* objects then I will change this to whatever they want, otherwise I will leave it the same and someone else can open a PR.

Reply:
Maybe just use `set`, etc, as a compromise? The operators aren't used anywhere else in the project I think

Reply:
I mean it looks fine (operator or otherwise), but relying on lenses to set a record seems overwrought.